numpy/distutils/misc_util.py

```diff
@@ -285,24 +291,22 @@ def gpaths(paths, local_path='', include_non_existing=True):
 
 
 def clean_up_temporary_directory():
-    tdata = tlocal()
-    _temporary_directory = getattr(tdata, 'tempdir', None)
-    if not _temporary_directory:
-        return
-    try:
-        shutil.rmtree(_temporary_directory)
-    except OSError:
-        pass
-    _temporary_directory = None
+    for d in _tmpdirs:
+        try:
+            shutil.rmtree(d)
+        except OSError:
+            pass
 
 def make_temp_file(suffix='', prefix='', text=True):
-    tdata = tlocal()
-    if not hasattr(tdata, 'tempdir'):
-        tdata.tempdir = tempfile.mkdtemp()
-        atexit.register(clean_up_temporary_directory)
+    if not hasattr(_tdata, 'tempdir'):
+        _tdata.tempdir = tempfile.mkdtemp()
+        with _tmpdirlock:
```

Comment:
Not sure the lock is really needed as append is thread safe. Not sure you need to hold the temporary directory in `_tdata.tempdir` either. That said, it probablly does not hurt.